src/gil.rs

```diff
@@ -370,6 +370,38 @@ extern "C" fn finalize() {
     }
 }
 
+/// Ensure the GIL is held, useful in implementation of APIs like PyErr::new where it's
+/// inconvenient to force the user to acquire the GIL.
+pub(crate) fn acquire_if_needed() -> GILAcquisition {
+    if gil_is_acquired() {
+        GILAcquisition::AlreadyHeld
+    } else {
+        GILAcquisition::JustInTime(GILGuard::acquire())
+    }
+}
+
+/// Enum used internally which avoids acquiring the GIL where it's not necessary.
+pub(crate) enum GILAcquisition {
```

Comment:
I like this abstraction since it helps to simplify the implementation  :+1: 
But I don't think we need naming each variant (AlreadyHeld` and `JustInTime`).
To make it simpler, how about  
```rust
pub(crate) struct EnsureGILHeld(Option<GILGuard>);
```
or so?

Reply:
:+1: I think struct is nicer than enum here - I will update PR.